packages/wrangler/src/pages/dev.ts

```diff
@@ -476,42 +521,130 @@ export const Handler = async (args: PagesDevArguments) => {
 					routesModule,
 					defineNavigatorUserAgent,
 				});
+
+				for (const dep in bundle.dependencies) {
+					const resolvedDep = resolve(functionsDirectory, dep);
+					/*
+					 * EXCLUDE:
+					 *   - the "/functions" directory because we're already watching it
+					 *   - everything in "./.wrangler", as it's mostly cache and
+					 *     temporary files
+					 *   - any bundle dependencies we are already watching
+					 *   - anything outside of the current working directory, since we
+					 *     are expecting `wrangler pages dev` to be run from the Pages
+					 *     project root folder
+					 */
+					if (
+						!resolvedDep.match(/\/functions\/./) &&
+						!resolvedDep.match(/\.wrangler\/./) &&
+						resolvedDep.match(resolve(process.cwd()))
+					) {
+						currentBundleDependencies.add(resolvedDep);
+
+						if (!watchedBundleDependencies.has(resolvedDep)) {
+							watchedBundleDependencies.add(resolvedDep);
+							watcher.add(resolvedDep);
+						}
+					}
+				}
+
+				// handle non-JS module dependencies, such as wasm/html/binary imports
+				for (const module of bundle.modules) {
+					if (module.filePath) {
+						const resolvedDep = resolve(functionsDirectory, module.filePath);
+						currentBundleDependencies.add(resolvedDep);
+
+						if (!watchedBundleDependencies.has(resolvedDep)) {
+							watchedBundleDependencies.add(resolvedDep);
+							watcher.add(resolvedDep);
+						}
+					}
+				}
+
+				/*
+				 *`bundle.dependencies` and `bundle.modules` will always contain the
+				 * latest dependency list of the current bundle. If we are currently
```

Comment:
soooo...turns out, while testing the `_worker.js` work, that there is at least one corner case where this seems not to be true.

Say our Function `math-is-fun.ts` imports `../math/add.ts` and calls `add()`. So far so good. Now imagine we delete the contents of `add.ts` but leave everything else as it is. We'll get a compile warning saying:

`Import "add" will always be undefined because the file "../math/add.ts" has no exports`

which is correct. At this point, IMHO, `add.ts` is technically still a dependency of `math-is-fun.ts` (though maybe we can argue abt that later), so my expectation would be that if we were to undo the changes we made in `add.ts`, that would trigger a rebuild. This works as expected if esbuild was in `watch` mode, but it does not trigger anything with the current implementation. The reason for that is that the `metafile` returned by esbuild does not list `add.ts` as an `input` to the entrypoint, which means we won't return it as part of `bundle.dependecies` 😕. The returned metafile BTW, looks exactly the same in both esbuild/chokidar watch modes. So, it's either that esbuild watches more files than what it returns in the metafile, or we're making some wrong assumptions wehn we process that metafile to return `bundle.dependencies`, or there is a plugin somewhere that strips out empty imports? (meaning of 0kb in size).

Whatever it is, the finding is a tad worrying...cause it makes me wonder...what other corner case I didn't think of yet, would result in a similar behaviour

it's never easy is it 😅 🍦 (`i-scream` 😉 )

Reply:
raised here: https://github.com/cloudflare/workers-sdk/issues/6182